File: ml-parcial-IIND3118/ml-parcial-IIND3118/temas/B13_manifold_learning/python/B13_manifold_learning.py
"""
B13. Manifold Learning: Isomap, LLE y t-SNE
Tema B13_manifold_learning — Introducción al Machine Learning (IIND3118)

CÓMO USARLO
  1. Cambia el bloque marcado con ✏️ abajo por tu dataset.
  2. Corre el archivo:  python B13_manifold_learning.py
  3. Las gráficas se guardan en la carpeta figuras/ de este tema.

La explicación del método está en el README.md de esta misma carpeta.
"""

import os, sys, warnings
warnings.filterwarnings("ignore")

import numpy as np
import pandas as pd
import matplotlib
matplotlib.use("Agg")          # no abre ventanas: funciona en cualquier entorno
import matplotlib.pyplot as plt

from sklearn.model_selection import train_test_split, KFold, cross_val_score, GridSearchCV
from sklearn.preprocessing import scale
from sklearn.linear_model import (LinearRegression, Ridge, Lasso, ElasticNet,
                                  RidgeCV, LassoCV, ElasticNetCV, LogisticRegression)
from sklearn.neighbors import KNeighborsRegressor, KNeighborsClassifier
from sklearn.metrics import (mean_squared_error, mean_absolute_error, r2_score,
                             accuracy_score, confusion_matrix, classification_report,
                             roc_curve, roc_auc_score)
from sklearn.decomposition import PCA, KernelPCA
from sklearn.cluster import KMeans, DBSCAN
from sklearn.ensemble import IsolationForest
from sklearn.neighbors import LocalOutlierFactor
from sklearn import manifold

# ── Guardado automático de figuras ──────────────────────────────────────
AQUI = os.path.dirname(os.path.abspath(__file__))
FIGS = os.path.join(AQUI, "..", "figuras")
os.makedirs(FIGS, exist_ok=True)
_n = [0]
def _save():
    _n[0] += 1
    ruta = os.path.join(FIGS, f"{_n[0]:02d}_figura.png")
    plt.savefig(ruta, dpi=110, bbox_inches="tight"); plt.close()
    print(f"   [figura guardada: figuras/{os.path.basename(ruta)}]")
plt.show = _save                          # cada plt.show() guarda en vez de abrir

RNG = np.random.default_rng(0)

# ╔═════════════════════════════════════════════════════════════════════════╗
# ║  ✏️  CAMBIA SOLO ESTE BLOQUE POR TU DATASET                             ║
# ╚═════════════════════════════════════════════════════════════════════════╝
DATOS  = os.path.join(AQUI, "..", "..", "..", "data", "ejemplo_regresion.csv")
TARGET = "target"          # nombre EXACTO de la columna respuesta. None si no hay.
# ╚═════════════════════════════════════════════════════════════════════════╝

df = pd.read_csv(DATOS).dropna()

if TARGET is not None:
    y_serie = df[TARGET]; X_df = df.drop(columns=[TARGET])
else:
    y_serie = None; X_df = df.copy()

X_df = pd.get_dummies(X_df, drop_first=True).astype(float)
nombres_X = list(X_df.columns)
X  = X_df.values
Xs = scale(X)

if y_serie is not None:
    y = pd.factorize(y_serie)[0] if y_serie.dtype == object else y_serie.values
    ES_CLASIFICACION = (len(np.unique(y)) <= 10) and np.all(y == np.round(y))
    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=.30, random_state=0)
    Xs_train, Xs_test, _, _ = train_test_split(Xs, y, test_size=.30, random_state=0)
else:
    y = None; ES_CLASIFICACION = False

def ModeloLineal(**kw):
    return (LogisticRegression(max_iter=5000, **kw) if ES_CLASIFICACION
            else LinearRegression(**kw))

col = y if y is not None else "steelblue"
Z   = PCA().fit_transform(Xs)        # proyección PCA, se usa para graficar
Z2  = Z[:, :2]

print("="*74)
print("B13. Manifold Learning: Isomap, LLE y t-SNE")
print("="*74)
print(f"datos: {X.shape[0]} obs. x {X.shape[1]} predictores")
if y is not None:
    print(f"y ('{TARGET}'): " + ("CLASIFICACIÓN" if ES_CLASIFICACION else "REGRESIÓN"))
print()

# ════════════════════════════════════════════════════════════════════════
#  CÓDIGO DEL MÉTODO
# ════════════════════════════════════════════════════════════════════════

# ═══ B13. Manifold Learning ═══

K_VEC = min(10, len(Xs)-1)       # ⬅ n_neighbors: el parámetro crítico
PERP  = min(30, (len(Xs)-1)//3)  # ⬅ perplexity de t-SNE (5-50)

emb_pca  = Z[:, :2]
emb_iso  = manifold.Isomap(n_neighbors=K_VEC, n_components=2).fit_transform(Xs)
# LLE es numéricamente delicado: si 'ltsa' falla, se prueban las otras
# variantes. 'hessian' es la más frágil de todas.
emb_lle = None
for metodo in ["ltsa", "modified", "standard"]:          # ⬅ variantes de LLE
    try:
        emb_lle = manifold.LocallyLinearEmbedding(
            n_neighbors=K_VEC, n_components=2,
            method=metodo, eigen_solver="dense",         # 'dense' es más estable
            random_state=0).fit_transform(Xs)
        print(f"LLE: funcionó con method='{metodo}'")
        break
    except Exception as e:
        print(f"LLE: method='{metodo}' falló ({type(e).__name__})")
if emb_lle is None:
    emb_lle = Z[:, :2]
    print("LLE falló con todas las variantes; se muestra PCA en su lugar.")
emb_tsne = manifold.TSNE(n_components=2, perplexity=PERP, init="pca",
                         random_state=0).fit_transform(Xs)

fig, ax = plt.subplots(1, 4, figsize=(17, 4))
for a,(e,t) in zip(ax, [(emb_pca,"PCA (lineal)"), (emb_iso,f"Isomap (k={K_VEC})"),
                        (emb_lle,"LLE"), (emb_tsne,f"t-SNE (perp={PERP})")]):
    a.scatter(e[:,0], e[:,1], c=col, cmap="viridis", s=20, alpha=.8)
    a.set_title(t); a.set_xticks([]); a.set_yticks([])
plt.tight_layout(); plt.show()

# ── Sensibilidad a k en Isomap ─────────────────────────────────────────
ks_try = [3, 5, 10, 25, min(60, len(Xs)-1)]
fig, ax = plt.subplots(1, len(ks_try), figsize=(3.2*len(ks_try), 3.4))
for a, kk in zip(ax, ks_try):
    try:
        e = manifold.Isomap(n_neighbors=kk, n_components=2).fit_transform(Xs)
        a.scatter(e[:,0], e[:,1], c=col, cmap="viridis", s=15)
        a.set_title(f"k={kk}")
    except Exception:
        a.set_title(f"k={kk}: falla\n(grafo desconectado)")
    a.set_xticks([]); a.set_yticks([])
fig.suptitle("Isomap: con k grande aparecen ATAJOS que rompen el manifold")
plt.tight_layout(); plt.show()

# ── ⚠️ EL CONTROL: ¿qué sale con datos SIN estructura? ────────────────
ruido = np.random.default_rng(0).standard_normal(Xs.shape)   # ruido puro
fig, ax = plt.subplots(1, 3, figsize=(14, 4))
ax[0].scatter(*PCA(n_components=2).fit_transform(ruido).T, s=15, alpha=.7)
ax[0].set_title("PCA sobre RUIDO")
for i, pp in enumerate([5, PERP]):
    e = manifold.TSNE(n_components=2, perplexity=pp, init="random",
                      random_state=0).fit_transform(ruido)
    ax[i+1].scatter(*e.T, s=15, alpha=.7)
    ax[i+1].set_title(f"t-SNE sobre RUIDO (perp={pp})")
for a in ax: a.set_xticks([]); a.set_yticks([])
fig.suptitle("Todos dibujan algo, aunque NO haya nada que encontrar", fontsize=13)
plt.tight_layout(); plt.show()
print("Con perplexity baja, t-SNE fabrica grumos convincentes sobre ruido puro.")
print("MORALEJA: un gráfico bonito NO es evidencia de estructura.")

print("\nListo. Revisa la carpeta figuras/")

File: ml-parcial-IIND3118/ml-parcial-IIND3118/temas/B13_manifold_learning/R/B13_manifold_learning.R
# ############################################################################
# B13. MANIFOLD LEARNING
# Tema B13_manifold_learning — Introducción al Machine Learning (IIND3118)
#
# CÓMO USARLO
#   1. Cambia el bloque marcado con ✏️ por tu dataset.
#   2. Corre:  Rscript B13_manifold_learning.R     (o ábrelo en RStudio)
#   3. Las gráficas se guardan en la carpeta figuras/ de este tema.
#
# La explicación del método está en el README.md de esta misma carpeta.
# ############################################################################

for (p in c("MASS","cluster","dbscan","glmnet","leaps","caret","class","fpc",
            "clusterCrit","isotree","kernlab","Rtsne","psych","car","vegan")) {
  if (!requireNamespace(p, quietly=TRUE))
    try(install.packages(p, repos="https://cloud.r-project.org"), silent=TRUE)
}
suppressMessages({ library(MASS); library(cluster) })

set.seed(0)

# Las gráficas se guardan en figuras/ en vez de abrirse en pantalla
dir.create("../figuras", showWarnings=FALSE)
.n <- 0
guardar <- function() {
  .n <<- .n + 1
  dev.copy(png, sprintf("../figuras/%02d_figura.png", .n), width=900, height=650)
  invisible(dev.off())
}

# ╔═══════════════════════════════════════════════════════════════════════╗
# ║  ✏️  CAMBIA SOLO ESTE BLOQUE POR TU DATASET                           ║
# ╚═══════════════════════════════════════════════════════════════════════╝
DATOS  <- "../../../data/ejemplo_regresion.csv"
TARGET <- "target"        # nombre EXACTO de la columna respuesta. NULL si no hay.
# ╚═══════════════════════════════════════════════════════════════════════╝

df <- na.omit(read.csv(DATOS))

if (!is.null(TARGET)) {
  y_raw <- df[[TARGET]]
  X_df  <- df[, setdiff(names(df), TARGET), drop=FALSE]
} else { y_raw <- NULL; X_df <- df }

X_df <- as.data.frame(model.matrix(~ . - 1, data=X_df))
nombres_X <- names(X_df)
X  <- as.matrix(X_df)
Xs <- scale(X)

if (!is.null(y_raw)) {
  ES_CLASIFICACION <- is.factor(y_raw) || is.character(y_raw) || length(unique(y_raw)) <= 10
  y <- if (ES_CLASIFICACION) as.factor(y_raw) else as.numeric(y_raw)
  idx <- sample(seq_len(nrow(X)), floor(0.70*nrow(X)))
  Xtrain <- X[idx,]; Xtest <- X[-idx,]; ytrain <- y[idx]; ytest <- y[-idx]
  Xs_train <- Xs[idx,]; Xs_test <- Xs[-idx,]
} else { y <- NULL; ES_CLASIFICACION <- FALSE }

pca  <- prcomp(X, center=TRUE, scale.=TRUE)
Z2   <- pca$x[, 1:2]
col_pt <- if (!is.null(y)) as.integer(as.factor(y)) else "steelblue"

cat(strrep("=",74), "\n")
cat("B13. MANIFOLD LEARNING\n")
cat(strrep("=",74), "\n")
cat(sprintf("datos: %d obs. x %d predictores\n\n", nrow(X), ncol(X)))

# ############################################################################
#  CÓDIGO DEL MÉTODO
# ############################################################################

# ============================================================================
# B13. MANIFOLD LEARNING (ISOMAP, LLE, t-SNE)
# ----------------------------------------------------------------------------
# LA IDEA: a veces los datos viven sobre una SUPERFICIE DOBLADA dentro del
# espacio. Imagina una hoja de papel enrollada en 3D: en realidad es 2D,
# solo está torcida. PCA la APLASTA y pone juntos puntos que sobre la hoja
# están lejísimos. Estos métodos la DESENROLLAN.
# Clave: distinguir la distancia EN LÍNEA RECTA (a través del aire) de la
# distancia CAMINANDO SOBRE LA SUPERFICIE (geodésica).
#
# ISOMAP = MDS CON DISTANCIAS GEODÉSICAS
#   Construye un grafo de k vecinos, calcula caminos más cortos (eso aproxima
#   la geodésica) y aplica MDS.
#   ✅ Preserva estructura GLOBAL
#   ❌ MUY sensible a k: k grande -> crea ATAJOS que saltan de un pliegue a
#      otro y destruyen todo; k chico -> el grafo se desconecta · O(n^3)
#
# LLE: preserva RELACIONES LINEALES LOCALES.
#   ❌ No preserva distancias globales · tiende a COLAPSAR regiones
#
# t-SNE: convierte distancias en PROBABILIDADES y minimiza la divergencia KL.
#   ⚠️ LO QUE NO SE PUEDE LEER DE UN t-SNE (esto lo preguntan)
#      - LOS TAMAÑOS DE LOS CLUSTERS NO SIGNIFICAN NADA
#      - LAS DISTANCIAS ENTRE CLUSTERS NO SIGNIFICAN NADA
#      - Puede INVENTAR clusters en datos sin estructura
#      - NO es determinista · NO proyecta datos nuevos
#
# PARÁMETROS
#   Isomap(k=)      : ⚠️ EL CRÍTICO. 5-15 típico
#   Rtsne(perplexity=) : 5-50. Baja -> grumos falsos; alta -> se aplana
#   Rtsne(dims=2, max_iter=)
#
# ⚠️ EL CONTROL OBLIGATORIO
#   Estos métodos SIEMPRE dibujan algo bonito, incluso con RUIDO PURO.
#   Antes de interpretar, corre el mismo método sobre datos permutados.
# ============================================================================

library(Rtsne)
K_VEC <- min(10, nrow(Xs)-1)
PERP  <- min(30, floor((nrow(Xs)-1)/3))

col_pt <- if(!is.null(y)) as.integer(as.factor(y)) else "steelblue"

# t-SNE
set.seed(0)
ts <- Rtsne(Xs, dims=2, perplexity=PERP,   # ⬅ perplexity: 5-50
            max_iter=500, check_duplicates=FALSE, verbose=FALSE)

# Isomap (RDRToolbox está en Bioconductor; vegan es la alternativa en CRAN)
emb_iso <- NULL
if (requireNamespace("vegan", quietly=TRUE)) {
  emb_iso <- vegan::isomap(dist(Xs), k=K_VEC, ndim=2)$points[,1:2]
} else if (requireNamespace("RDRToolbox", quietly=TRUE)) {
  emb_iso <- RDRToolbox::Isomap(data=Xs, dims=2, k=K_VEC)[[1]]
}

par(mfrow=c(1,3))
plot(pca$x[,1:2], col=col_pt, pch=16, main="PCA (lineal)", xlab="", ylab="")
if (!is.null(emb_iso))
  plot(emb_iso, col=col_pt, pch=16, main=sprintf("Isomap (k=%d)", K_VEC), xlab="", ylab="")
else plot.new(); title("Isomap: instala vegan")
plot(ts$Y, col=col_pt, pch=16, main=sprintf("t-SNE (perp=%d)", PERP), xlab="", ylab="")
par(mfrow=c(1,1))

# ⚠️ EL CONTROL: ¿qué sale con datos SIN estructura?
set.seed(0)
ruido <- matrix(rnorm(nrow(Xs)*ncol(Xs)), nrow=nrow(Xs))
ts_r  <- Rtsne(ruido, dims=2, perplexity=5, check_duplicates=FALSE, verbose=FALSE)

par(mfrow=c(1,2))
plot(prcomp(ruido)$x[,1:2], pch=16, cex=.6, main="PCA sobre RUIDO", xlab="", ylab="")
plot(ts_r$Y, pch=16, cex=.6, main="t-SNE sobre RUIDO (perp=5)", xlab="", ylab="")
par(mfrow=c(1,1))
cat("Con perplexity baja, t-SNE FABRICA grumos convincentes sobre ruido puro.\n")
cat("MORALEJA: un gráfico bonito NO es evidencia de estructura.\n")


# ############################################################################
# TABLA FINAL — QUÉ MÉTODO USAR
# ############################################################################
#
# CON VARIABLE RESPUESTA y (SUPERVISADO)
#   Punto de partida siempre              -> Regresión lineal      (A7)
#   Relación no lineal, pocas variables   -> KNN / kernel          (A3)
#   Muchas variables correlacionadas      -> RIDGE                 (A9)
#   Muchas variables, pocas importan      -> LASSO (da ceros)      (A9)
#   Correlación Y quieres selección       -> ELASTIC NET           (A9)
#   p pequeño, quieres el mejor subconj.  -> Exhaustivo            (A8)
#   p grande, seleccionar rápido          -> Forward               (A8)
#   Comparar sin partir la muestra        -> AIC/BIC/Cp/R2adj      (A5)
#   Calibrar cualquier parámetro          -> Validación cruzada    (A6)
#
# SIN y (NO SUPERVISADO)
#   Agrupar, clusters redondos, n grande  -> K-Means               (B1)
#   Agrupar con outliers o categóricas    -> K-Medoides + Gower    (B2)
#   Ver jerarquía, no fijar K             -> Jerárquico            (B3)
#   Formas raras + detectar ruido         -> DBSCAN                (B4)
#   Elegir K                              -> Codo / Silhouette     (B5)
#   Outliers, pocas variables             -> Mahalanobis           (B6)
#   Outliers, muchas variables            -> Isolation Forest      (B7)
#   Outliers locales                      -> LOF                   (B7)
#   Reducir dimensión, visualizar         -> PCA                   (B8)
#   Solo tengo distancias                 -> MDS                   (B9)
#   Factores latentes interpretables      -> Análisis Factorial    (B10)
#   No lineal + proyectar datos nuevos    -> Kernel PCA            (B11)
#   Comprimir matriz / LSI                -> SVD                   (B12)
#   Solo visualizar estructura curva      -> Isomap / t-SNE        (B13)
#
#
# ############################################################################
# ⚠️ LAS 15 TRAMPAS DEL CURSO
# ############################################################################
#  1. El MSE de TRAIN siempre baja con más flexibilidad. Calibra con test o CV.
#  2. La curva del MSE de TEST tiene forma de U. El óptimo es el mínimo.
#  3. En KNN, Ridge y kernel el PARÁMETRO GRANDE = MODELO RÍGIDO (al revés).
#  4. BIC castiga más que AIC (p*log n vs 2p) => BIC elige modelos más chicos.
#  5. Accuracy ENGAÑA con clases desbalanceadas. Mira la matriz de confusión.
#  6. Lasso da CEROS exactos (selecciona); Ridge NUNCA. La diferencia clave.
#  7. HAY QUE ESTANDARIZAR en KNN, Ridge/Lasso, K-Means, PCA, DBSCAN.
#  8. Forward NO garantiza el mejor modelo (heurística); el exhaustivo sí.
#  9. K-Means SIEMPRE devuelve K grupos, aunque no haya estructura.
# 10. Silhouette/CH/DB premian a K-Means sobre DBSCAN en anillos concéntricos.
# 11. El Z-Score se ENMASCARA con grupos de outliers. Usa Z modificado.
# 12. Mirar variable por variable NO ES Mahalanobis (rompe la correlación).
# 13. p-valor ALTO en análisis factorial = modelo ADECUADO (al revés).
# 14. En t-SNE, tamaños y distancias ENTRE clusters NO significan nada.
# 15. Manifold learning SIEMPRE dibuja algo bonito, aunque sea ruido puro.
#
# ############################################################################
# 🔗 EQUIVALENCIAS QUE PREGUNTAN
# ############################################################################
#   MDS clásico (euclidiano)        = PCA
#   PCA                             = SVD sobre datos centrados
#   Isomap                          = MDS con distancias geodésicas
#   Kernel PCA con kernel lineal    = PCA
#   min WSS                         = max BSS  (ANOVA, solo K-Means euclidiano)
#   Elastic Net alpha=0 = Ridge  |  alpha=1 = Lasso
#
# ############################################################################
# DIFERENCIAS R vs PYTHON (para no confundirse)
# ############################################################################
#   Ruido en DBSCAN     R: 0            Python: -1
#   LOF                 R: mayor=atípico  Python: negative_outlier_factor_ al revés
#   Isolation Forest    R: mayor=anómalo  Python: decision_function al revés
#   SVD                 R: devuelve v     Python: devuelve Vt (transpuesta)
#   Ward                R: "ward.D2"      Python: "ward"
#   Reinicios K-Means   R: nstart         Python: n_init
#   Índices             R: desde 1        Python: desde 0
# ############################################################################

cat("\nListo. Revisa la carpeta figuras/\n")
